compiler/src/dmd/lexer.d

```diff
@@ -2023,7 +2023,12 @@ class Lexer
                     error(loc, "%d does not fit in 16 bits", d2);
                 u = d1;
                 if (d2)
-                    u = (d1 << 16) | d2;
+                {
+                    static if (1)
+                        u = d2; // emulate gcc behavior
+                    else
+                        u = (d1 << 16) | d2;
+                }
```

Comment:
It would be better to understand why gcc does this for `u'ab'`, according to the spec (6.4.4.4-9), these are `char16_t` literals, so it doesn't seem to be correct to ignore the top half of the literal.


| Prefix | Corresponding Type |
| ------  | ------------------------ |
| none | unsigned char |
| L | the unsigned type corresponding to wchar_t |
| u | char16_t |
| U | char32_t |

Reply:
Looking at the grammar.
```
character-constant:
    encoding-prefix(opt) ' c-char-sequence '

encoding-prefix:
    u8
    u
    U
    L

c-char-sequence:
    c-char
    c-char-sequence c-char
```

So why are we limiting 16-bit character literals to just two sequences?  Ditto 32-bit character literals to just one?

Having more than one character is obviously nonsense, as only the last character in the sequence ultimately becomes the value in both cases.  But as it's only a semantic warning, why reject it?

Reply:
It's implementation defined. I simply adjusted it so it did what gcc does. ImportC doesn't do warnings, it just accepts it.